src/compile/data/optimize.ts

```diff
@@ -199,20 +207,26 @@ function optimizationDataflowHelper(dataComponent: DataComponent) {
 export function optimizeDataflow(data: DataComponent) {
   // check before optimizations
   checkLinks(vals(data.sources));
-
-  for (let i = 0; i < 5; i++) {
+  let warningFlag = false;
+  for (let i = 0; i < MAX_OPTIMIZATION_RUNS; i++) {
     if (!optimizationDataflowHelper(data)) {
       break;
     }
+    if (i === MAX_OPTIMIZATION_RUNS - 1) {
+      warningFlag = true;
+    }
   }
 
   // move facets down and make a copy of the subtree so that we can have scales at the top level
   vals(data.sources).map(moveFacetDown);
 
-  for (let i = 0; i < 5; i++) {
+  for (let i = 0; i < MAX_OPTIMIZATION_RUNS; i++) {
     if (!optimizationDataflowHelper(data)) {
       break;
     }
+    if (warningFlag || i === MAX_OPTIMIZATION_RUNS - 1) {
```

Comment:
This is weird. Let's just add one counter variable and don't initialize it in the for loop. Okay?